### PR TITLE
feat: implement group invitation system; auto-accept pending invitations on user signup; add tests for invitation creation and handling

### LINKED ISSUE #29: Invite a User to a Group
### Acceptance Criteria
- Given I am on a group's member page, When I enter a valid email address and click "Invite", Then an invitation is created in the system and sent to that email.

### Definition of Done
- Code complete, passes all tests, reviewed by a peer and merged into the main branch.

### Dependencies / Risks
- Depends on core group functionality being in place. Also depends on the email sending service.

### PR DESCRIPTION
**Pull request by PradHolla**
## 📋 Description
Invite User to a group and auto add them once they sign up with the application

## 🎯 Changes Made
- Added `GroupInvitation` model to track pending invitations
- Modified group creation to send invitations for non-existent users
- Added auto-accept logic when invited users sign up
- Created console notification for group invitations

## 🧪 Testing
- [x] All tests pass (`uv run pytest -v`)
- [x] Coverage meets requirements (85%+ per file, 90%+ overall)
- [x] Added 7 new tests in `test_group_invitations.py`
- [x] Updated existing tests to match new behavior

## 🔗 Related Issue
Closes #29 

## 📸 Screenshots

User invite on the console:

Screenshot 2025-10-27 161622

Since the invited user hasn't signed up yet, we don't see their email in the member list:

Screenshot 2025-10-27 161632

After signing up, their email is displayed:

Screenshot 2025-10-27 161730

Screenshot 2025-10-27 161754

## 🔄 Files Changed
- `models.py` - Added GroupInvitation model
- `routes/groups.py` - Modified create_group() for invitations
- `routes/auth.py` - Added auto-accept on signup
- `services/notification_service.py` - Added notify_group_invitation()
- `tests/test_group_invitations.py` - New test file (7 tests)
- `tests/test_group_routes.py` - Updated existing tests
- `tests/test_expense_routes.py` - Updated existing tests

## 📝 Notes
- Changed from auto-creating users to invitation system
- Invited users cannot be in expenses until they sign up
- Duplicate invitations are prevented
- Email validation done before creating invitations

### CODE REVIEW

File: routes/groups.py [hunk 1]
```diff
@@ -67,15 +68,31 @@ def create_group():
...
+                        )
+                        db.session.add(invitation)
+
+                        # Send invitation notification
+                        try:
+                            notify_group_invitation(creator.email, email, name)
+                        except Exception as e:
+                            print(f"Failed to send invitation notification: {e}")
 
         db.session.commit()
```
[the remark this answers is not in the document]
Comment (PradHolla) [reply routes/groups.py:101]: added db.session.flush()

File: routes/groups.py [hunk 2]
```diff
@@ -67,15 +68,31 @@ def create_group():
...
+                        invitation = GroupInvitation(
+                            email=email, group_id=group.id, invited_by_id=user_id
+                        )
+                        db.session.add(invitation)
+
+                        # Send invitation notification
+                        try:
+                            notify_group_invitation(creator.email, email, name)
+                        except Exception as e:
+                            print(f"Failed to send invitation notification: {e}")
```
[the remark this answers is not in the document]
Comment (PradHolla) [reply routes/groups.py:99]: Will do this later

**Review by sairithik9849 — APPROVED** [review]
## Summary

This PR implements a robust group invitation system with auto-acceptance on signup, along with comprehensive test coverage. The changes effectively address the requirement for inviting users who haven't yet signed up to the platform.

## Strengths

**1. Clear Implementation**
- The `GroupInvitation` model is well-structured with appropriate fields (email, group_id, invited_by_id, status, created_at)
- The logic clearly separates existing users (added immediately) from non-existent users (invitation created)
- Auto-accept on signup (in `verify_otp`) is implemented cleanly with a proper check to avoid duplicate memberships

**2. Excellent Test Coverage**
- New test file `test_group_invitations.py` with 338 lines provides comprehensive coverage
- Tests cover happy paths, edge cases (duplicate invitations, user already in group), error handling (notification failures), and invalid inputs
- Integration with auth flow tested in `test_verify_otp_auto_accepts_pending_invitations`
- Existing tests properly updated to reflect new behavior

**3. Good Developer Experience**
- Notification service provides clear feedback (printed to terminal for now)
- Duplicate invitation prevention avoids database clutter
- Error handling with try-catch around notifications ensures robustness

## Suggestions for Improvement

**1. Database Considerations**
- Consider adding a database index on `(email, group_id, status)` since this combination is frequently queried (lines 85-87 in groups.py, lines 86 in auth.py)
- Consider adding a unique constraint on `(email, group_id, status='pending')` at the database level to prevent race conditions

**2. Edge Cases & Maintainability**
- The `db.session.flush()` on line 65 of groups.py is critical but could benefit from a more detailed comment explaining the race condition it prevents
- Consider what happens to invitations when a group is deleted - should they cascade delete? Add `ondelete='CASCADE'` to the foreign key if appropriate
- The status field accepts string values - consider using an Enum or at least document valid values clearly (currently only in comment on line 112)
- What happens if a user declines an invitation? The status supports 'declined' but there's no endpoint to handle this action yet

**3. Potential Improvements**
- Consider adding an expiration mechanism for invitations (e.g., expire after 30 days)
- The notification failure is silently caught with just a print statement - consider logging this properly for production monitoring
- Consider sending a confirmation notification to the inviter when someone accepts their invitation
- Line 97 in groups.py: email invitations would be more useful than terminal prints for production use

**4. Minor Code Quality**
- Line 99 in groups.py: bare `except Exception` is quite broad - consider catching specific exceptions
- Consider extracting the auto-accept logic from `verify_otp` into a separate service method for better testability and separation of concerns

## Security & Data Integrity

✅ Properly validates email format before creating invitations
✅ Checks for existing invitations to prevent duplicates
✅ Verifies user is not already a member before adding
✅ Uses proper foreign key relationships

## Overall Assessment

This is a well-implemented feature with strong test coverage. The code is clean, follows existing patterns, and handles most edge cases appropriately. The suggestions above are primarily for long-term maintainability and production readiness rather than blocking issues.

**Recommendation:** Approve with minor suggestions for future enhancements.
